stdlib/public/core/NormalizedCodeUnitIterator.swift

```diff
@@ -226,15 +226,96 @@ extension _NormalizedUTF8CodeUnitIterator: Sequence { }
 internal
 struct _NormalizedUTF16CodeUnitIterator: IteratorProtocol {
   internal typealias CodeUnit = UInt16
-  var segmentBuffer = _FixedArray16<CodeUnit>(allZeros:())
-  var overflowBuffer: [CodeUnit]? = nil
-  var normalizationBuffer: [CodeUnit]? = nil
+  var segmentBuffer = NormalizationBuffer()
+  var intermediateBuffer = NormalizationBuffer()
   var source: _SegmentSource
-
-  var segmentBufferIndex = 0
-  var segmentBufferCount = 0
-  var overflowBufferIndex = 0
-  var overflowBufferCount = 0
+  
+  internal struct NormalizationBuffer {  
+    enum Mode {
+      case stack, heap
+    }
+    var stackStorage: _FixedArray16<CodeUnit>
+    var heapStorage: [CodeUnit] = []
+    var count: Int
+    var readIndex: Int
+    var mode: Mode = .stack
+  
+    init() {
+      stackStorage = _FixedArray16(allZeros: ())
+      count = 0
+      readIndex = 0
+    }
+    
+    mutating func reset() {
+      count = 0
+      readIndex = 0
+    }
+    
+    var isStackBased: Bool {
+      return mode == .stack
+    }
+    
+    var canRead: Bool {
+      return readIndex < count
+    }
+    
+    mutating func read() -> CodeUnit {
+      defer { readIndex += 1 }
+      return self[readIndex]
+    }
+    
+    mutating func promote(_ size: Int, multiplier: Int = 1) {
+      if isStackBased {
+        heapStorage = [CodeUnit](repeating: 0, count: (size + count) * multiplier)
+        for i in 0..<count {
+          heapStorage[i] = stackStorage[i]
+        }
+        mode = .heap
+      } else {
+        fatalError("Promoting a heap array doesn't make sense")
+      }
+    }
+    
+    subscript(index: Int) -> CodeUnit {
+      assert(index < count)
+      assert(index >= 0)
```

Comment:
These should be `_sanityCheck`s, otherwise the stdlib has to always compiles them regardless of user compilation mode.